src/nvim/tui/tui.c

```diff
@@ -348,50 +370,210 @@ static void update_attrs(UI *ui, HlAttrs attrs)
       data->params[1].i = (fg >> 8) & 0xff;   // green
       data->params[2].i = fg & 0xff;          // blue
       unibi_out(ui, data->unibi_ext.set_rgb_foreground);
+      data->default_attr = false;
     }
 
     if (bg != -1) {
       data->params[0].i = (bg >> 16) & 0xff;  // red
       data->params[1].i = (bg >> 8) & 0xff;   // green
       data->params[2].i = bg & 0xff;          // blue
       unibi_out(ui, data->unibi_ext.set_rgb_background);
+      data->default_attr = false;
     }
   } else {
     if (fg != -1) {
       data->params[0].i = fg;
       unibi_out(ui, unibi_set_a_foreground);
+      data->default_attr = false;
     }
 
     if (bg != -1) {
       data->params[0].i = bg;
       unibi_out(ui, unibi_set_a_background);
+      data->default_attr = false;
     }
   }
 
   if (attrs.bold) {
     unibi_out(ui, unibi_enter_bold_mode);
+    data->default_attr = false;
   }
   if (attrs.italic) {
     unibi_out(ui, unibi_enter_italics_mode);
+    data->default_attr = false;
   }
   if (attrs.underline || attrs.undercurl) {
     unibi_out(ui, unibi_enter_underline_mode);
+    data->default_attr = false;
   }
   if (attrs.reverse) {
     unibi_out(ui, unibi_enter_reverse_mode);
+    data->default_attr = false;
   }
 }
 
 static void print_cell(UI *ui, UCell *ptr)
 {
+  TUIData *data = ui->data;
+  UGrid *grid = &data->grid;
+  if (data->immediate_wrap_after_last_column
+      && grid->row >= ui->height - 1
+      && grid->col >= ui->width - 1) {
+    // This (rare) kind of terminal simply cannot print in this corner without
+    // scrolling the entire screen up a line, which we do not want to happen.
+    return;
+  }
   update_attrs(ui, ptr->attrs);
   out(ui, ptr->data, strlen(ptr->data));
+  ++grid->col;
+}
+
+static bool cheap_to_print(UI *ui, int row, int col, int next)
+{
+  TUIData *data = ui->data;
+  UGrid *grid = &data->grid;
+  UCell *cell = grid->cells[row] + col;
+  while (next) {
+    --next;
```

Comment:
If you don't mind, please use post-inc/dec. Else I'll have to fix it before merging, to avoid failing the lint check. We decided on a style and we stick to it.